eventbroadcaster/notifiersink.go

```diff
@@ -29,34 +27,48 @@ type NotifierSink struct {
 	logger   *zap.SugaredLogger
 }
 
-// RegisterNotifierSinks builds notifiers sinks and registers them on the given broadcaster
-func RegisterNotifierSinks(mgr ctrl.Manager, broadcaster record.EventBroadcaster, notifiersConfig eventnotifier.NotifiersConfig, logger *zap.SugaredLogger) (err error) {
-	client := mgr.GetClient()
-
-	notifiers, err := eventnotifier.GetNotifiers(notifiersConfig, logger)
-
+// RegisterNotifierSinks registers notifiers sinks on the given broadcaster
+func RegisterNotifierSinks(mgr ctrl.Manager, broadcaster record.EventBroadcaster, notifiers []eventnotifier.Notifier, logger *zap.SugaredLogger) {
 	for _, notifier := range notifiers {
 		logger.Infof("notifier %s enabled", notifier.GetNotifierName())
 
-		broadcaster.StartRecordingToSink(&NotifierSink{client: client, notifier: notifier, logger: logger})
+		broadcaster.StartRecordingToSink(&NotifierSink{client: mgr.GetClient(), notifier: notifier, logger: logger})
 	}
 
 	corev1Client, _ := corev1client.NewForConfig(mgr.GetConfig())
 
 	broadcaster.StartRecordingToSink(&corev1client.EventSinkImpl{Interface: corev1Client.Events("")})
-
-	return
 }
 
 func (s *NotifierSink) Create(event *corev1.Event) (*corev1.Event, error) {
-	dis, err := s.getDisruption(event)
-	if err != nil {
+	s.logger.Infof("CREATE event received: %s", event.Message)
+
+	var obj client.Object
+
+	switch event.InvolvedObject.Kind {
+	case v1beta1.DisruptionKind:
+		disruption, err := s.getDisruption(event)
+		if err != nil {
+			s.logger.Warn(err)
+			return event, fmt.Errorf("eventnotifier: unable to get Disruption object")
```

Comment:
Here I have changed the behavior by returning an error rather `nil`, to retry the event. If we return `nil` the event will not be retried and it will increase the chance to miss events.

Reply:
in what kind of situation would it suceed again after an error? How many times does it try to retry?

Reply:
The number of retry is hardcoded in the k8s library:
- https://github.com/DataDog/chaos-controller/blob/main/vendor/k8s.io/client-go/tools/record/event.go#L39
- https://github.com/DataDog/chaos-controller/blob/main/vendor/k8s.io/client-go/tools/record/event.go#L293

Reply:
> in what kind of situation would it suceed again after an error?

The event can be emitted  and handled before the creation of the resource. In this case, the k8s client will not find the resource and returns an error. So in order to succeed it should retry until the resource is found.

Reply:
Could we mention somewhere in a comment that automatic retrying is handled by client-go up to 12 times?